### ISSUE TITLE
Should IntersectionObserver => target be a bijection?

### ISSUE THREAD

**Issue by szager-chromium** [slightlyoff/IntersectionObserver#81]
It has been suggested that the API would be more ergonomic if an IntersectionObserver represented an observation between a single root and a single target:

new IntersectionObserver(callback, targetElement, optionDict);

This would get rid of the observe() and unobserve() methods; disconnect() would remain.

Pros:

- More intuitive
- Requires less global bookkeeping

Cons:

- Less memory efficient (parameters like rootMargin are stored per-target, rather than per-observer).
- Callbacks would not be batched.  Currently, if an IntersectionObserver has multiple observed targets that generate notifications within a single cycle, the callback is invoked once with a sequence of notification records.  With the change, there would be one callback per target.

This is a brief summary of an email conversation; please add details.  The general sentiment is that the current spec is not a deal-breaker, but perhaps not ideal.  In practice, developers are free to use one observer per target, and shouldn't suffer any negative consequences.

An open question is whether, in practice, batching callbacks for multiple targets of a single observer is actually useful and more performant.  It's hard to know for certain until we see what developers do with it.

**Comment by bmaurer** [comment]
My perspective as a site implementer -- the current spec makes it really unclear if it is important that a site batch using as few observers as possible. It's possible that different implementations will have different performance constraints.

Grouping elements by the parameters (such as rootMargin) requires an implementer to create a wrapper around the API. An API could either offer a fixed number of observers (here's one that has no margin, etc) which is brittle or have a cache of observers (requiring a fair amount of code). Alternatively the callers could make no effort to batch at all in which case InteractionObserver is just an overly verbose API.

If nothing else, it would be nice to have the spec clarify that browsers should expect to be able to support a high number of observers

**Comment by szager-chromium** [comment]
You know, the difference between one-target-per-observer and many-targets-per-observer really comes down to the explicit calls to observe() and unobserve().  But for that, it's possible to simply ignore the difference.  So maybe one way to bridge the gap here is to allow an optional 'target' argument to the IntersectionObserver constructor, which would do an implicit observe(target); this:

var observer = new IntersectionObserver(callback, target, optionDict);

... would be equivalent to:

var observer = new IntersectionObserver(callback, optionDict);
observer.observe(target);

To create an observer without a first target, but with options, would be:

new IntersectionObserver(callback, null, optionDict);

The spec would specify this behavior, and call out the potential benefits of reusing an observer if it's expected to generate multiple notifications during a single frame.  Other than that, the spec probably shouldn't touch on possible performance advantages, since that would be implementation-specific.

How does that sound?

**Comment by szager-chromium** [comment]
Tab points out to me that's possible to write the IntersectionObserver constructor IDL with overloads to handle two- and three-argument cases:

new IntersectionObserver(callback, target [, optionDict]);
new IntersectionObserver(callback [, optionDict]);

**Comment by mpb** [comment]
+1
Sounds pretty good, actually.

**Comment by esprehn** [comment]
I don't think we should do that, it diverges the API from MutationObserver and PerformanceObserver. Calling .observe() is not very hard or verbose.

**Comment by szager-chromium** [comment]
What if we make the observe() method return the IntersectionObserver object?  Then at least it could be written as a one-liner:

var observer = (new IntersectionObserver(callback, optionDict)).observe(target);

**Comment by esprehn** [comment]
Note that this has the same performance characteristics as MutationObserver. You can use a single observer for many elements, or multiple observers, both should perform similarly except for the extra wrapper and allocation cost, but I would hope you're not having hundreds of observers in your page.

How many do you anticipate needing @bmaurer ? The design of the API is not what's going to make this slow, the implementation has lots of things that make having tons of observers "slow", but I'd rather see it for real in a page first before messing with it.

**Comment by esprehn** [comment]
I'd rather we didn't make .observe() any different than MutationObserver. :) I don't think making this a one liner is of huge value, your code to process the callback or the dictionary is going to be a lot more than one line, saving the assignment to a local variable isn't a lot of work.

**Comment by bmaurer** [comment]
> How many do you anticipate needing @bmaurer?

Right now in FB's codebase there are over 100 callsites to APIs that we would want to replace with IntersectionObserver. My concern is that in a complex codebase it's hard to get people to agree on just a few IntersectionObservers because that implies finding settings (like rootMargin) that work for everybody. Given the complexity of our codebase we'd probably do one of two things:

1) Create an IntersectionObserver factory. Callers would pass in parameters like rootMargin and we'd batch them up into a minimal number of observers.

2) Just create one IntersectionObserver each time we want to observe an element (even if they have the same parameters)

(1) is fairly complex. If minimizing the number of IntersectionObservers is important than I feel like many sites will have to create a similar factory. For (2) my concern is that some browsers might assume that sites will not use many observers and will make that pattern have poor performance.

I'm not too concerned with the verbosity of the current spec -- we can always create helper functions. My concern is that the current API forces a user to either write a IntersectionObserverFactory or run the risk that browsers might be far less efficient at handling many observers.

**Comment by ojanvafai** [comment]
There are two sides to this discussion:
1. Developer ergonomics of the API surface. This comes down to the two options szager listed in  https://github.com/slightlyoff/IntersectionObserver/issues/81#issuecomment-173335693. I suspect that authors using this API directly will mostly be happy as is and that framework authors will mostly make wrappers. That sounds fine to me. Hopefully we'll get some more data points as we reach out to more authors.

2. Performance overhead of having many observers. The potential overhead for many observers would be a  detail of our implementation and not affected by which of the two options we pick. On the other hand, there's significant performance improvement that comes with having a single callback that takes a list of change records for all the elements involved instead of one callback per element (e.g. don't need to thunk back and forth between JS and C++ per callback). In practice, I expect that performance benefit will play out similarly across browsers.

Given those two, I'm inclined to say we not change this part of the API unless we get more developer feedback that this will be painful. 

Ben, please clarify if I mischaracterized. Particularly, if you do experiment with this, I'd love to see any performance data you encounter. I suspect that the performance overhead will be dominated by the number of elements you are observing, not by the number of observers you make, which is not affected by the shape of API we expose.

**Comment by ojanvafai** [comment]
Sorry, I raced with Ben on that last comment. Ben, TL;DR: I wouldn't worry too much about creating many observers to start with. It will only matter if you're observing a large number of elements, but then you'll have bigger problems (or...we will since we'll need to rework our implementation).

**Comment by ojanvafai** [comment]
My intuition is that the main perf overhead of many observers is actually that you'd have many callbacks, which is, I think, explicitly what you want. I belive that if we change the API to be more callback-per-observe based, it would have a similar level of overhead to the observer-per-observe approach.

**Comment by ojanvafai** [comment]
We should add advice to the explainer here for this and make sure to cover this subtlety clearly in any developer documentation we write for this feature. Mainly, that if you have multiple targets with the same root, you will get better performance if you use a single IntersectionObserver.

**Comment by m59peacemaker** [comment]
If I understood the concerns here, then I think we can separate them into two concepts:
  - we want an API that has per-target listeners without having to create additional observers
  - we want to reuse existing observers based, which is dependent upon the arguments used to construct them

I believe I solved the first concern with [`insular-observer`](https://www.npmjs.com/package/insular-observer). Suggestions welcome!

The second concern I think is solved by a `memoize` function that can handle arguments such as are used to construct IntersectionObserver. I'm not sure how to memoize a function that takes an object containing a DOM node, but I think there are npm packages that can do it.

So, I think finding the right `memoize` function and then:

```js
const makeAnObserveFunction = memoize(InsularObserver)

// make an `observe` function that has a nice API reusing existing observers if they match the type/options 
const observeSomething = makeAnObserveFunction(IntersectionObserver, yourOptions)

const stopObserving = observeSomething(target, listener)
```

And then you would just use your `makeAnObserveFunction` everywhere you need any kind of Observer in a project, and you will create the minimum needed observers have a simple API for using them. What do you all think?

**Comment by XPD-Kasun** [comment]
Is there any update here?
